ADVANCED_EXAMPLES.md

```diff
@@ -0,0 +1,596 @@
+This is a set of handy **composable** stamps, as well as few tips and tricks.
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Self aware objects - `instance.getStamp()`. 2 ways.](#self-aware-objects---instancegetstamp-2-ways)
+  - [Attach function to object](#attach-function-to-object)
+  - [Attach function to prototype (memeory efficient)](#attach-function-to-prototype-memeory-efficient)
+- [Self cloneable objects - `instance.clone()`. 3 ways.](#self-cloneable-objects---instanceclone-3-ways)
+  - [Attach method to each object instance](#attach-method-to-each-object-instance)
+  - [Bind method to each object instance](#bind-method-to-each-object-instance)
+  - [Attach function to prototype (memory efficient)](#attach-function-to-prototype-memory-efficient)
+- [Delayed object instantiation using Promises](#delayed-object-instantiation-using-promises)
+- [Dependency injection tips](#dependency-injection-tips)
+- [Validate before a function call](#validate-before-a-function-call)
+- [EventEmitter without inheritance (`convertConstructor`)](#eventemitter-without-inheritance-convertconstructor)
+- [Mocking in the unit tests](#mocking-in-the-unit-tests)
+- [Hacking stamps](#hacking-stamps)
+  - [Enforced default properties](#enforced-default-properties)
+  - [Enforced default properties as a composable behavior](#enforced-default-properties-as-a-composable-behavior)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+
+## Self aware objects - `instance.getStamp()`. 2 ways.
+
+> Run the examples below for yourself:
+```sh
+$ git clone https://github.com/stampit-org/stampit.git
+$ cd stampit && npm i babel -g
+$ babel-node advanced-examples/self-aware.js
+```
+
+You can add `.getStamp()` function to each object with ease. 
+
+First, let's assume you have a following stamp:
+```js
+const User = stampit();
+```
+
+### Attach function to object
+Just compose the following stamp to any other stamp.
+```js
+const SelfAware1 = stampit.init(({ instance, stamp }) => {
+  instance.getStamp = () => stamp;
+});
+```
+Let's compose it with the `User` stamp from above:
+```js
+const SelfAwareUser1 = User.compose(SelfAware1);
```

Comment:
I just meant that the variables should be renamed in this file, `ADVANCED_EXAMPLES.md`. This file is what explains the advanced use cases, it should be as clear as possible. The example files can be left as is I think.